### PR TITLE
Client Telemetry: Adds memory optimization by reducing Histogram significant digits

### LINKED ISSUE #2881: Client Telemetry: Histogram Memory Management
**Problem Statement:**

[HDRHistogram library](http://www.hdrhistogram.org/) has been used for Client Telemetry calculation but looks like it allocates too much memory (below table) with current configuration in .Net which needs to be optimized. 

**Current Configuration and Memory Allocations :**
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">
<head>
<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/SOURAB~1/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/SOURAB~1/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#0563C1" vlink="#954F72">

Metrics |   | Histogram   Allocation | Comments
-- | -- | -- | --
CPU | new   LongConcurrentHistogram(1, 99999, 5); | 2 MB | Expecting   histogram to have Minimum CPU Usage of .001% and Maximum CPU Usage of 999.99%
Request   Latency | new   LongConcurrentHistogram(1, TimeSpan.TicksPerHour, 5); | 20 MB | Expecting   histogram to have Minimum Latency of 1 and Maximum Latency of 1 hour
Request   Charge | new   LongConcurrentHistogram(1, 9999900, 5); | 8 MB | Expecting   histogram to have Minimum Request Charge of 1 and Maximum Request Charge of   9999900     Maximum precision can be set as 5 so consider maximum value in 5 digit i.e.   99999 of request charge     So 99999 * HistogramPrecisionFactor = 9999900 is the maximum request charge   we have set.     For all the Document ReadWriteQuery Operations there will be at least 1   request charge
Memory | new   LongConcurrentHistogram(1, Int64.MaxValue, 5); | 48 MB | Expecting   histogram to have Minimum Memory Remaining of 1 MB and Maximum Memory   Remaining of Long Max Value
</body>
</html>

**Analysis and Current Implementation**
Same library (Java version) has been used in Java Client Telemetry for same calculation. There are below difference between Java and .Net:
1. In Java, we have `ConcurrentDoubleHistogram `which records `double `values and .Net we have `ConcurrentLongHistogram `which records `long `values. ( I hope it won't make any difference , I checked memory allocation wise, it was quite similar).
due to which in .Net SDK, we are multiplying all the values by 100 to save precision of data.
2. In Java, `numberOfSignificantDigits` is set as 2 which means value quantization will be within range upto 1/100th (i.e. 1%).

**Open Questions**
1. Should we worry about the data precision while recording values in histogram? Anyway, there is no guarantee that Histogram will record the same exact value.
2. With `numberOfSignificantDigits` as 2 configuration, value quantization will be within range upto 1/100th (i.e. 1%). If understand it correctly, any value can change upto 1%. Are we fine with it?

### PR DESCRIPTION
**Pull request by sourabh1007**
## Description
Optimized Memory Allocation while creating or assigning histogram. Made it same as Java.

Metrics | Old   | New | Improved
-- | -- | -- | --
CPU | 2,000 KB | 63 KB | -96.85%
Request Latency | 20,000 KB | 2996 KB | -85%
Request Charge | 8,000 KB | 70 KB | -99.1%
Memory | 48,000 KB | 110 KB |  -99.8%

## Type of change
Number requests 2881

- [] Bug fix (non-breaking change which fixes an issue) i.e 12.5%

## Closing issues

To automatically close an issue: closes #2881

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Telemetry/ClientTelemetryOptions.cs [hunk 1]
```diff
@@ -33,21 +33,21 @@ internal static class ClientTelemetryOptions
...
-        internal const int RequestChargePrecision = 5;
+        internal const int RequestChargePrecision = 2;
         internal const string RequestChargeName = "RequestCharge";
         internal const string RequestChargeUnit = "RU";
 
         // Expecting histogram to have Minimum CPU Usage of .001% and Maximum CPU Usage of 999.99%
         internal const long CpuMax = 99999;
         internal const long CpuMin = 1;
-        internal const int CpuPrecision = 5; // 100 (max CPU Usage) * 100(Historam Precision factor) = 10000 which is 5 digit and that is what we can set as maximum precision in histogram
+        internal const int CpuPrecision = 2; // 100 (max CPU Usage) * 100(Historam Precision factor) = 10000 which is 5 digit and that is what we can set as maximum precision in histogram
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Telemetry/ClientTelemetryOptions.cs:43]: nit: does the comment need to be updated?

**Review by ealsur — APPROVED** [review]
LGTM

File: Microsoft.Azure.Cosmos/src/Telemetry/ClientTelemetryOptions.cs [hunk 2]
```diff
@@ -23,31 +23,29 @@ internal static class ClientTelemetryOptions
...
-        internal const int CpuPrecision = 5; // 100 (max CPU Usage) * 100(Historam Precision factor) = 10000 which is 5 digit and that is what we can set as maximum precision in histogram
+        internal const int CpuPrecision = 2; 
         internal const String CpuName = "CPU";
         internal const String CpuUnit = "Percentage";
 
         // Expecting histogram to have Minimum Memory Remaining of 1 MB and Maximum Memory Remaining of Long Max Value
         internal const long MemoryMax = Int64.MaxValue;
         internal const long MemoryMin = 1;
-        internal const int MemoryPrecision = 5;
+        internal const int MemoryPrecision = 2;
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Telemetry/ClientTelemetryOptions.cs:48]: Can we add tests to validate this to prevent regressions?

Comment (sourabh1007) [reply Microsoft.Azure.Cosmos/src/Telemetry/ClientTelemetryOptions.cs:48]: It does n't change the data precision (anyway we are storing number in "long"). It changes how histogram assign buckets and hence memory allocation.

**Pull request merged**